### PR TITLE
feat(conditions): add updated-at condition

### PR DESCRIPTION
**Pull request by sileht**
Change-Id: I8bdf4821605d3b9efcbf473197294095c3066d02

### CODE REVIEW

**Comment by sileht** [comment]
This pull request is part of a stack:
1. feat(rules): allow and/or in conditions ([#2691](https://github.com/Mergifyio/mergify-engine/pull/2691))
1. feat: allow merged/closed/author/base/head in and/or conditions ([#2692](https://github.com/Mergifyio/mergify-engine/pull/2692))
1. feat(conditions): allow checks condition in and/or ([#2693](https://github.com/Mergifyio/mergify-engine/pull/2693))
1. refactor(filter): make it generic ([#2694](https://github.com/Mergifyio/mergify-engine/pull/2694))
1. feat(time): add time conditions ([#2695](https://github.com/Mergifyio/mergify-engine/pull/2695))
1. feat(conditions): add day/month/year/day-of-week conditions ([#2705](https://github.com/Mergifyio/mergify-engine/pull/2705))
1. feat(conditions): add support for named day-of-week ([#2708](https://github.com/Mergifyio/mergify-engine/pull/2708))
1. feat(conditions): add calendar condition ([#2709](https://github.com/Mergifyio/mergify-engine/pull/2709))
1. feat(conditions): add update-at condition ([#2710](https://github.com/Mergifyio/mergify-engine/pull/2710))

**Comment by sileht** [comment]
test

File: mergify_engine/tests/unit/rules/test_parser.py [hunk 1]
```diff
@@ -128,6 +132,27 @@
...
+            "updated-at>=18:02",
+            {">=": ("updated-at", now + datetime.timedelta(hours=18, minutes=2))},
+        ),
+        (
+            f"updated-at<={now.isoformat()}",
+            {"<=": ("updated-at", now)},
+        ),
+        (
+            "updated-at<=7 days",
+            {"<=": ("updated-at", now + datetime.timedelta(days=7))},
```
Comment (jd) [inline mergify_engine/tests/unit/rules/test_parser.py:145]:
```suggestion
            "updated-at<=7 days",
            {"<=": ("updated-at", now + datetime.timedelta(days=7))},
```

I don't think this is right. It means it gets translated to `update-at <= now + 7 days`.

If you write `updated-at<=7 days` you mean: `updated-at` is between now and 7 days ago which means:
`now-7days<=update-at<now()`. The `<now()` is obviously true, so we can skip it so it boils down to `now -  7 days <= updated-at` which is equivalent to `now <= updated_at + 7 days`.

**Review by jd — CHANGES_REQUESTED** [review]

File: mergify_engine/date.py [hunk 2]
```diff
@@ -66,3 +67,13 @@ def from_string(cls, string: str) -> "DayOfWeek":
 
     def __str__(self) -> str:
         return self._SHORT_DAY[self.value - 1].capitalize()
+
+
+def fromisoformat(s: str) -> datetime.datetime:
```
[the remark this answers is not in the document]
Comment (sileht) [reply mergify_engine/date.py:72]: For 1. I have plan to move `utcnow()` to the  module `date`. I will do that separatly.

File: mergify_engine/tests/unit/rules/test_parser.py (hunk 1 above)
Comment (sileht) [reply mergify_engine/tests/unit/rules/test_parser.py]: :facepalm: 
```patch
- +
+ - 
```

Comment (jd) [reply mergify_engine/tests/unit/rules/test_parser.py]: Actually, it's not just changing the sign. We're not talking about the same things.

1. When I write `updated-at <= 7 days` I mean "updated-at is less than 7 days old" which means: `now - 7 days <= updated-at`

2. When I wrote `updated-at <= 7 days ago` I mean "updated-at date is before the day it was 7 days ago" which translates to `updated-at <= now - 7 days`.

The problem here is that the language says 1 but the code runs 2.

We might need 2 different syntaxes here to make things very clear. It's not clear to me.

**Review by jd — CHANGES_REQUESTED** [review]
unclear

**Pull request closed without merging**